File: Programmers/Level_3/여행경로.py
'''
https://programmers.co.kr/learn/courses/30/lessons/43164

stack으로 DFS 하는 방법 
'''

from collections import defaultdict

def solution(tickets):
    answer = []

    # 예정지 목록 
    d = defaultdict(list)

    for item in tickets:
        d[item[0]].append(item[1])

    # 오름차순으로 정렬하면 pop 연산 대신 remove 연산을 해야돼서 O(n) 시간이 걸리니까
    # 내림차순으로 정렬해서 pop 연산을 통해 사전 순으로 가장 앞에 있는걸 O(1) 시간으로 제거 
    for key in d:
        d[key].sort(reverse=True)

    # 목적지 목록 
    stack = ['ICN']

    while stack:
        # 다음 목적지 
        top = stack[-1]

        # 만약 목적지에서 또 다른 공항으로 넘어갈 수 없다면 
        # 더이상 이 경로를 통해 갈 수 있는 곳이 없으니 정답에 추가 
        if not d[top]:
            answer.append(stack.pop())
        
        # 목적지 목록에 추가하고 예정지에선 제거 
        else:
            stack.append(d[top].pop())

    # 위에서 내림차순 정렬을 했었으니 다시 오름차순으로 정렬해서 리턴 
    return list(reversed(answer))

File: Programmers/Level_3/Swift_Code/Swift_Code/여행경로.swift
//
//  여행경로.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/05/26.
//

/*
 https://programmers.co.kr/learn/courses/30/lessons/43164?language=swift
 
 출발지를 key, 목적지를 value로 취급하여 hash table을 만든다
 value를 내림차순으로 정렬한다 (pop으로 사전순으로 가장 앞에 있는 것 부터 방문)
 스택에 방문하는 노드를 넣고 그 노드에서 갈 수 있는 사전 순으로 가장 앞서는 도느 찾기
 만약 더 이상 갈 수 있는 노드가 없다면 방문한 노드를 답에 추가 
 */

import Foundation

func solution(_ tickets:[[String]]) -> [String] {
    var answer = [String]()
    var d = [String: [String]]()
    var stack = [String]()
    
    for ticket in tickets {
        d[ticket[0], default: [String]()].append(ticket[1])
    }
    
    for key in d.keys {
        d[key]?.sort(by: >)
    }
    
    stack.append("ICN")
    
    while !stack.isEmpty {
        let top = stack.last!
        
        if d[top, default: [String]()].count == 0 {
            answer.append(stack.removeLast())
        } else {
            stack.append(d[top]!.removeLast())
        }
    }
    
    return answer.reversed()
}

// print(solution([["ICN", "JFK"], ["HND", "IAD"], ["JFK", "HND"]]))
